src/main/scala/cromwell/binding/Scope.scala

```diff
@@ -1,14 +1,168 @@
 package cromwell.binding
 
+import cromwell.binding.AstTools.AstNodeName
+import cromwell.parser.WdlParser.{AstList, Ast, AstNode}
+import scala.collection.JavaConverters._
+
+
+object Scope {
+
+  /**
+   * Utility class to generate a ScopeTree from an Ast.
+   */
+  private class ScopeTreeGenerator {
+
+    /* Keep track of the # of created Scatter nodes. */
+    var scatterCount: Int = 0
```

Comment:
Yes so like I said in one of Miguel's comments, I completely agree that this doesn't look great but I actually think that it's not as bad as it seems.

Because the function that build the "Tree" is recursive it makes it quite difficult to track the number of created scatters along the way, especially for "sibling scatters". Like Miguel suggested it can probably be done by returning the index at every step, but I see 2 drawbacks in doing this:
- It complexifies the return type of the function : I guess something like a pair of (Int, Option[Scope]), or even a temporary object containing both.
It'll also be more and more complicated to handle as we add new scopes that need indexing.
- The function itself will be less readable as there will be more work to extract the scopes and index from the function response, and feed them in the right way to subsequent calls.

While here I think it's really clear what this var does and also its scope is extremely limited.
Actually I also just realized that the class itself is not necessary, everything can be put inside a function.

Anyway I'm still gonna spend some time to try to find a better immutable way to do this without having a mess of indexes.